lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2039,9 +2039,16 @@ Type TypeChecker::typeCheckExpression(Expr *&expr, DeclContext *dc,
   if (options.contains(TypeCheckExprFlags::AllowUnresolvedTypeVariables))
     allowFreeTypeVariables = FreeTypeVariableBinding::UnresolvedType;
 
+  Type convertTo = convertType.getType();
+  if (options.contains(TypeCheckExprFlags::ExpressionTypeMustBeOptional)) {
+    assert(!convertTo && "convertType and type check options conflict");
+    Type var = cs.createTypeVariable(cs.getConstraintLocator(expr));
```

Comment:
It would be great if you could add a new `ConstraintLocator::PathElementKind` for this, so it would be possible to identify such "contextual hints" based on their locators.